Trouble creating a ZIP archive with lots of entries

I am chasing an issue that I initially saw when testing XrdEc. I have since come up with a standalone testing case that is not related to using XrdEc. The test code and instruction to compile the code is pasted at the bottom. Note that

1. Once compile, pls run against a xrootd server to created a new ZIP archive
2. The program will never finish. If turning on XRD_LOGEVEL=Error, I see message like the following after a while:
[2023-04-30 21:13:01.694793 -0700][Error  ][AsyncSock         ] [cent7b:1094.0] Socket error encountered: [ERROR] Socket error: connection reset by peer
[2023-04-30 21:13:01.694852 -0700][Error  ][XRootD            ] [cent7b:1094] Impossible to send message kXR_writev (handle: 0x00000000, chunks: 1, total size: 2097211). Trying to recover.
[2023-04-30 21:13:01.694863 -0700][Error  ][XRootD            ] [cent7b:1094] Unable to get the response to request kXR_writev (handle: 0x00000000, chunks: 1, total size: 2097211)

3. it can not be compiled with devel rpm because of "#include <XrdCl/XrdClCtx.hh>". This is what is used in XrdEc.
4. This issue is sensitive to the setting of MAX_FILES_IN_ZIP, size_t fsize and std::string fn in the testing code. If we reduce MAX_FILES_IN_ZIP by 1 (while keeping "fsize" and "fn" unchanged), all will work just fine. 
5. It is possible that the MAX_FILES_IN_ZIP setting is related to my environment. You may need to set it to something much larger to see this issue. 

---------------------------------------

/* How to compile
 scl enable devtoolset-9 sh
 g++ -g -o xrdzip_write.exe xrdzip_write.cc \
     -I/afs/slac/package/xrootd/xrootd-5.5.1/xrootd/src \
     -I/afs/slac/package/xrootd/xrootd-5.5.1/amd64_rhel70/src \
     -Wl,--copy-dt-needed-entries \
     -L/afs/slac/package/xrootd/xrootd-5.5.1/amd64_rhel70/src \
     -L/afs/slac/package/xrootd/xrootd-5.5.1/amd64_rhel70/src/XrdCl \
     -l XrdCl  
*/

#include <stdio.h>
#include <string>
#include <XrdCl/XrdClXRootDResponses.hh>
#include <XrdCl/XrdClURL.hh>
#include <XrdCl/XrdClFile.hh>
#include <XrdCl/XrdClCtx.hh>
#include <XrdCl/XrdClFileSystem.hh>
#include <XrdCl/XrdClZipArchive.hh>
#include <XrdCl/XrdClZipOperations.hh>
#include <XrdOuc/XrdOucCRC32C.hh>

#define MAX_FILES_IN_ZIP 36251

int main(int argc, char* argv[]) {
    XrdCl::URL url(argv[1]);
    XrdCl::XRootDStatus st;
    
    size_t fsize = 128*1024;
    char *buff = (char*)malloc(fsize);
    uint32_t cksm = crc32c(0, buff, fsize);

    XrdCl::ZipArchive zip;
    XrdCl::Ctx<XrdCl::ZipArchive> pzip = zip;

    st = XrdCl::WaitFor(XrdCl::OpenArchive(pzip, url.GetURL(), XrdCl::OpenFlags::New | XrdCl::OpenFlags::Update));
    if (!st.IsOK()) {
        printf("ZipArchive Open fail\n");
        exit(1);
    }
    else
        printf("ZipArchive Open successfully\n");

    for (int i=0; i<MAX_FILES_IN_ZIP; i++) {
        std::string fn = "obj." + std::to_string(i) + ".1";
        printf("Open file %s : ", fn.c_str());
        st = XrdCl::WaitFor(XrdCl::AppendFile(pzip, fn, cksm, fsize, buff));
        if (!st.IsOK()) {
           printf("fail %s\n", st.ToString().c_str());
           exit(1);
        }
        else
           printf("successfully\n");
    }

    printf("closing ZipArchive ...\n");
    st = XrdCl::WaitFor(XrdCl::CloseArchive(pzip));
    if (!st.IsOK()) {
        printf("ZipArchive Close fail\n");
        exit(1);
    }
    else
        printf("ZipArchive Close successfully\n");

    exit(0);
}

--- a/src/XrdCl/XrdClOperations.cc
+++ b/src/XrdCl/XrdClOperations.cc
@@ -210,6 +210,32 @@ namespace XrdCl
     final = std::move( f );
   }
 
+  //------------------------------------------------------------------------
+  // Called by a pipeline on the handler of its first operation before Run
+  //------------------------------------------------------------------------
+  void PipelineHandler::PreparePipelineStart()
+  {
+    // Move any final-function from the handler of the last operaiton to the
+    // first. It will be moved along the pipeline of handlers while the
+    // pipeline is run.
+
+    if( final || !nextOperation ) return;
+    PipelineHandler *last = nextOperation->handler.get();
+    while( last )
+    {
+      Operation<true> *nextop = last->nextOperation.get();
+      if( !nextop ) break;
+      last = nextop->handler.get();
+    }
+    if( last )
+    {
+      // swap-then-move rather than only move as we need to guarantee that
+      // last->final is left without target.
+      std::function<void(const XRootDStatus&)> f;
+      f.swap( last->final );
+      Assign( std::move( f ) );
+    }
+  }
 
   //------------------------------------------------------------------------
   // Stop the current pipeline

--- a/src/XrdCl/XrdClOperations.hh
+++ b/src/XrdCl/XrdClOperations.hh
@@ -123,6 +123,11 @@ namespace XrdCl
       //------------------------------------------------------------------------
       void Assign( std::function<void(const XRootDStatus&)>  final );
 
+      //------------------------------------------------------------------------
+      //! Called by a pipeline on the handler of its first operation before Run
+      //------------------------------------------------------------------------
+      void PreparePipelineStart();
+
     private:
 
       //------------------------------------------------------------------------
@@ -487,6 +492,10 @@ namespace XrdCl
         if( !operation ) std::logic_error( "Empty pipeline!" );
 
         Operation<true> *opr = operation.release();
+        PipelineHandler *h = opr->handler.get();
+        if( h )
+          h->PreparePipelineStart();
+
         opr->Run( timeout, std::move( prms ), std::move( final ) );
       }
 

--- a/src/XrdCl/XrdClUtils.cc
+++ b/src/XrdCl/XrdClUtils.cc
@@ -865,4 +865,54 @@ namespace XrdCl
       if( dst_supported.count( *itr ) ) return *itr;
     return std::string();
   }
+
+  //----------------------------------------------------------------------------
+  //! Split chunks in a ChunkList into one or more ChunkLists
+  //----------------------------------------------------------------------------
+  void Utils::SplitChunks( std::vector<ChunkList> &listsvec,
+                           const ChunkList        &chunks,
+                           const uint32_t          maxcs,
+                           const size_t            maxc )
+  {
+    listsvec.clear();
+    if( !chunks.size() ) return;
+
+    listsvec.emplace_back();
+    ChunkList *c    = &listsvec.back();
+    const size_t cs = chunks.size();
+    size_t idx      = 0;
+    size_t nc       = 0;
+    ChunkInfo tmpc;
+
+    c->reserve( cs );
+
+    while( idx < cs )
+    {
+      if( maxc && nc >= maxc )
+      {
+        listsvec.emplace_back();
+        c = &listsvec.back();
+        c->reserve( cs - idx );
+        nc = 0;
+      }
+
+      if( tmpc.length == 0 )
+        tmpc = chunks[idx];
+
+      if( maxcs && tmpc.length > maxcs )
+      {
+        c->emplace_back( tmpc.offset, maxcs, tmpc.buffer );
+        tmpc.offset += maxcs;
+        tmpc.length -= maxcs;
+        tmpc.buffer  = static_cast<char*>( tmpc.buffer ) + maxcs;
+      }
+      else
+      {
+        c->emplace_back( tmpc.offset, tmpc.length, tmpc.buffer );
+        tmpc.length = 0;
+        ++idx;
+      }
+      ++nc;
+    }
+  }
 }

--- a/src/XrdCl/XrdClUtils.hh
+++ b/src/XrdCl/XrdClUtils.hh
@@ -272,6 +272,18 @@ namespace XrdCl
         if( !st.IsOK() ) return false;
         return protver >= kXR_PROTPGRWVERSION;
       }
+
+      //------------------------------------------------------------------------
+      //! Split chunks in a ChunkList into one or more ChunkLists
+      //! @param listsvec        : output vector of ChunkLists
+      //! @param chunks          : input ChunkLisits
+      //! @param maxcs           : maximum size of a ChunkInfo in output
+      //! @param maxc            : maximum number of ChunkInfo in each ChunkList
+      //------------------------------------------------------------------------
+      static void SplitChunks( std::vector<ChunkList> &listsvec,
+                               const ChunkList        &chunks,
+                               const uint32_t          maxcs,
+                               const size_t            maxc );
   };
 
   //----------------------------------------------------------------------------

--- a/src/XrdCl/XrdClZipArchive.cc
+++ b/src/XrdCl/XrdClZipArchive.cc
@@ -28,6 +28,7 @@
 #include "XrdCl/XrdClLog.hh"
 #include "XrdCl/XrdClDefaultEnv.hh"
 #include "XrdCl/XrdClConstants.hh"
+#include "XrdCl/XrdClUtils.hh"
 #include "XrdZip/XrdZipZIP64EOCDL.hh"
 
 #include <sys/stat.h>
@@ -626,10 +627,18 @@ namespace XrdCl
       }
 
       auto wrtbuff = std::make_shared<buffer_t>( GetCD() );
-      chunks.emplace_back( cdoff, wrtbuff->size(), wrtbuff->data() );
+      Pipeline p = XrdCl::Write( archive, cdoff,
+                                 wrtbuff->size(),
+                                 wrtbuff->data() );
       wrtbufs.emplace_back( std::move( wrtbuff ) );
 
-      Pipeline p = XrdCl::VectorWrite( archive, chunks );
+      std::vector<ChunkList> listsvec;
+      XrdCl::Utils::SplitChunks( listsvec, chunks, 262144, 1024 );
+
+      for(auto itr = listsvec.rbegin(); itr != listsvec.rend(); ++itr)
+      {
+        p = XrdCl::VectorWrite( archive, *itr ) | p;
+      }
       if( ckpinit )
         p       |= XrdCl::Checkpoint( archive, ChkPtCode::COMMIT );
       p         |= Close( archive ) >>

